### PR TITLE
feat: add regular expressions in route

### PR DESCRIPTION
**Pull request by jorbriib**
The user is been able to add a regular expression to validate a dynamic route.

1. Lexer is in charge of capturing the regular expression 
2. Parser is in charge of compiling the regular expression
3. Tree is in charge of matching the parameter with the regular expression

### Refactors 
Lexer code split in a new file
Refactor to simplify the parser

### CODE REVIEW

File: lexer.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
+package http_router
```
Comment (sangarbe) [inline lexer.go:1]: much better having the lexer on its own file, you anticipated ;)

File: parser.go [hunk 2]
```diff
@@ -181,22 +53,31 @@ func (p *Parser) parseVar() (bool, error) {
...
-		return false, fmt.Errorf("Parser error, expected %s but got %s", "var identifier", token.v)
+		return false, fmt.Errorf("parser error, expected %s but got %s", "var identifier", token.v)
 	}
 	p.buf.Write([]byte(token.v))
 
 	token = p.lexer.scan()
+
+	var regExp *regexp.Regexp
+	if isRegExpressionToken(token) {
+		regExp = regexp.MustCompile(fmt.Sprintf("^%s$", token.v))
```
Comment (sangarbe) [inline parser.go:64]: good point with the start/ending of the regex <3

File: parser_test.go [hunk 3]
```diff
@@ -275,6 +78,23 @@ func TestParserDoesNotValidateInvalidPaths(t *testing.T) {
 	}
 }
 
+func TestParserThrowsPanicWhenExpressionIsInvalid(t *testing.T) {
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf("The code did not panic")
```
Comment (sangarbe) [inline parser_test.go:84]: good catch, but this makes me feel like refactor asap the panics and move to errors. I will prioritise this task

File: parser_test.go [hunk 4]
```diff
@@ -292,24 +112,32 @@ func TestParserChechChunks(t *testing.T) {
...
+		"[{0 / } {1 id } {0 / }]",
+		"[{0 / } {1 id } {0 /path1 }]",
+		"[{0 /asdf- } {1 id }]",
+		"[{0 / } {1 id } {0 -adfasf }]",
+		"[{0 / } {1 id } {0 / } {1 name }]",
+		"[{0 / } {1 id } {0 / } {1 name } {0 / }]",
+		"[{0 / } {1 id } {0 - } {1 name } {0 / }]",
+		"[{0 / } {1 id ^[0-9]+$} {0 / }]",
+		"[{0 /path1/ } {1 id ^/ab+c/$}]",
+		"[{0 /path1/ } {1 id ^/d(b+)d/g$} {0 /path2 }]",
```
Comment (sangarbe) [inline parser_test.go]: Are you sure this is what we expect regarding the start/ending of the regular expresion? I mean, the "/g" is a modifier isn't it? So I think the correct regex should be /^d(b+)d$/g. Have you checked this kind of regular expresion against some string? Let's talk about this

File: parser_test.go (hunk 3 above)
Comment (jorbriib) [reply parser_test.go:84]: Yes, you are right

**Review by sangarbe — APPROVED** [review]

**Pull request merged**